rp2040-hal/src/gpio/func.rs

```diff
@@ -0,0 +1,185 @@
+pub trait Function: func_sealed::Function {}
+
+/// Value-level `enum` for pin function.
+#[allow(missing_docs)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum DynFunction {
+    Xip,
+    Spi,
+    Uart,
+    I2c,
+    Pwm,
+    Sio(DynSioConfig),
+    Pio0,
+    Pio1,
+    Clock,
+    Usb,
+    Null,
+}
+
+/// Value-level `enum` for SIO configuration.
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum DynSioConfig {
+    /// Pin is configured as Input.
+    Input,
+    /// Pin is configured as Output.
+    Output,
+}
+
+impl Function for DynFunction {}
+impl func_sealed::Function for DynFunction {
+    fn from(f: DynFunction) -> Self {
+        f
+    }
+
+    fn as_dyn(&self) -> DynFunction {
+        *self
+    }
+}
+
+macro_rules! pin_func {
+    ($($fn:ident $(as $alias:ident)?),*) => {
+        $(paste! {
+            /// Type-level `variant` for pin [`Function`].
+            pub struct [<Function $fn>](pub(super) ());
+            impl Function for [<Function $fn>] {}
+            impl func_sealed::TypeLevelFunction for [<Function $fn>] {}
+            impl func_sealed::Function for [<Function $fn>] {
+                fn from(_f: DynFunction) -> Self {
+                    Self(())
+                }
+                fn as_dyn(&self) -> DynFunction {
+                    DynFunction::[<$fn>]
+                }
+            }
+            $(
+                #[doc = "Alias to [`Function" $fn "`]."]
+                pub type [<Function $alias>] = [<Function $fn>];
+            )?
+        })*
+    };
+}
+pin_func!(Xip, Spi, Uart, I2c as I2C, Pwm, Pio0, Pio1, Clock, Usb, Null);
+
+// =============================
+// SIO sub-types
+
+/// Type-level `variant` for pin [`Function`].
+pub struct FunctionSio<C>(PhantomData<C>);
+impl<C: SioConfig> Function for FunctionSio<C> {}
+impl<C: SioConfig> func_sealed::TypeLevelFunction for FunctionSio<C> {}
+impl<C: SioConfig> func_sealed::Function for FunctionSio<C> {
+    fn from(_f: DynFunction) -> Self {
+        FunctionSio(PhantomData)
+    }
+    fn as_dyn(&self) -> DynFunction {
+        DynFunction::Sio(C::DYN)
+    }
+}
+/// Alias to [`FunctionSio<Input>`].
+pub type FunctionSioInput = FunctionSio<SioInput>;
+/// Alias to [`FunctionSio<Output>`].
+pub type FunctionSioOutput = FunctionSio<SioOutput>;
+
+/// Type-level `enum` for SIO configuration.
+pub trait SioConfig {
+    #[allow(missing_docs)]
+    const DYN: DynSioConfig;
+}
+
+/// Type-level `variant` for SIO configuration.
+pub enum SioInput {}
+impl SioConfig for SioInput {
+    #[allow(missing_docs)]
+    const DYN: DynSioConfig = DynSioConfig::Input;
+}
+/// Type-level `variant` for SIO configuration.
+pub enum SioOutput {}
+impl SioConfig for SioOutput {
+    #[allow(missing_docs)]
+    const DYN: DynSioConfig = DynSioConfig::Output;
+}
+
+// =============================
```

Comment:
I like these kinds of dividers, especially in large complicated files. Happy to see them made even bigger :)